### PR TITLE
MergeTree `Client` Legacy API Removed

### PR DESCRIPTION
**Pull request by anthony-murphy**
The `Client` class in the merge-tree package has been removed. Additionally, types that directly or indirectly expose the merge-tree `Client` class have also been removed.

The removed types were not meant to be used directly, and direct usage was not supported:

- AttributionPolicy
- IClientEvents
- IMergeTreeAttributionOptions
- SharedSegmentSequence
- SharedStringClass

Some classes that referenced the `Client` class have been transitioned to interfaces. Direct instantiation of these classes was not supported or necessary for any supported scenario, so the change to an interface should not impact usage. This applies to the following types:

- SequenceInterval
- SequenceEvent
- SequenceDeltaEvent
- SequenceMaintenanceEvent

The initial deprecations of the now changed or removed types were announced in Fluid Framework v2.4.0:
[Several MergeTree Client Legacy APIs are now deprecated](https://github.com/microsoft/FluidFramework/blob/main/RELEASE_NOTES/2.4.0.md#several-mergetree-client-legacy-apis-are-now-deprecated-22629)

### CODE REVIEW

**Comment by anthony-murphy** [comment]
#22827 https://github.com/microsoft/FluidFramework/issues/22827

File: packages/dds/merge-tree/api-report/merge-tree.legacy.alpha.api.md
Comment (jason-ha) [inline packages/dds/merge-tree/api-report/merge-tree.legacy.alpha.api.md:405]: Caution - this API was not marked `@deprecated`. Removable because no longer transitively accessible?

File: packages/dds/sequence/package.json
Comment (jason-ha) [inline packages/dds/sequence/package.json:196]: Why full stop disable? Isn't this exactly the case where we want to list the incompatibilities? Is another option to change entrypoint from "legacy" to "public"?

File: packages/dds/sequence/src/intervalCollection.ts
Comment (jason-ha) [inline packages/dds/sequence/src/intervalCollection.ts:71]: What is preventing this from following preferred structure and being in ./intervals/index.js?
If keeping, please comment on the lint disable.

File: packages/dds/sequence/package.json
Comment (anthony-murphy) [reply packages/dds/sequence/package.json:196]: i don't want to deal with merge conflict right now. once we are closer to the merge point i'll remove it and deal with them

File: packages/dds/sequence/src/intervals/sequenceInterval.ts
Comment (jason-ha) [inline packages/dds/sequence/src/intervals/sequenceInterval.ts:48]: I thought the sorting would still want `type IIntervalHelpers` first.
Other pattern that has no such confusion is to have separate `import type {}` block. We do that more now.

File: packages/dds/sequence/src/intervals/sequenceInterval.ts
Comment (jason-ha) [inline packages/dds/sequence/src/intervals/sequenceInterval.ts:120]: This should have an explicit tsdoc and `@internal` tag. No complaints from the tooling?

File: packages/dds/sequence/src/intervalCollection.ts
Comment (jason-ha) [reply packages/dds/sequence/src/intervalCollection.ts:71]: x several more instances of this.

**Review by jason-ha — COMMENTED** [review]
Shouldn't you be able to make many of the test changes before the breaking change?
When we have a well staged deprecation, I would expect to see more removal of things we don't have to test for anymore compared to test adaptation. The latter suggests customers need to adapt too. OR we don't have good coverage on the external facing APIs which also hinders confidence.

File: packages/dds/sequence/src/intervals/sequenceInterval.ts
Comment (anthony-murphy) [reply packages/dds/sequence/src/intervals/sequenceInterval.ts:120]: no it shouldn't, it's never exported from the package

File: packages/dds/sequence/src/intervals/sequenceInterval.ts
Comment (anthony-murphy) [reply packages/dds/sequence/src/intervals/sequenceInterval.ts:48]: i don't know. i just let the formatter do what it wants.

File: packages/dds/sequence/src/intervalCollection.ts
Comment (anthony-murphy) [reply packages/dds/sequence/src/intervalCollection.ts:71]: sure. this isn't meant to be checkin ready. hence the draft

**Comment by anthony-murphy** [comment]
> Shouldn't you be able to make many of the test changes before the breaking change? When we have a well staged deprecation, I would expect to see more removal of things we don't have to test for anymore compared to test adaptation. The latter suggests customers need to adapt too. OR we don't have good coverage on the external facing APIs which also hinders confidence.

that is not possible given the changes. the changes literally rename the class and make it un-exported

**Comment by jason-ha** [comment]
Just having another staging thought. For the cases where you now have a FooClass and Foo is interface, I think you can currently do
`const FooClass = Foo;`
exporting as needed for internal use. Then there is better clarity where you need to preserve the class internally versus interface and less churn when the time comes.
Unfortunately, I don't know that we can get away with the reverse (`const Foo = FooClass;`) at least not without more song and dance per api-extractor.

**Comment by anthony-murphy** [comment]
> Just having another staging thought. For the cases where you now have a FooClass and Foo is interface, I think you can currently do `const FooClass = Foo;` exporting as needed for internal use. Then there is better clarity where you need to preserve the class internally versus interface and less churn when the time comes. Unfortunately, I don't know that we can get away with the reverse (`const Foo = FooClass;`) at least not without more song and dance per api-extractor.

i don't understand this comment. i'm not exporting anything concrete. classes are being replaces with interfaces, nothing but types are exported

**Comment by jason-ha** [comment]
> > Shouldn't you be able to make many of the test changes before the breaking change? When we have a well staged deprecation, I would expect to see more removal of things we don't have to test for anymore compared to test adaptation. The latter suggests customers need to adapt too. OR we don't have good coverage on the external facing APIs which also hinders confidence.
> 
> that is not possible given the changes. the changes literally rename the class and make it un-exported

I am not sure I follow. Either you are doing non-"public" testing and still internally "export"ing the class or you shouldn't need to access the class beyond its instance type (no ctor) for testing.

**Comment by anthony-murphy** [comment]
> > > Shouldn't you be able to make many of the test changes before the breaking change? When we have a well staged deprecation, I would expect to see more removal of things we don't have to test for anymore compared to test adaptation. The latter suggests customers need to adapt too. OR we don't have good coverage on the external facing APIs which also hinders confidence.
> > 
> > 
> > that is not possible given the changes. the changes literally rename the class and make it un-exported
> 
> I am not sure I follow. Either you are doing non-"public" testing and still internally "export"ing the class or you shouldn't need to access the class beyond its instance type (no ctor) for testing.

the test's are for implementations, they unit test the actual class which is only ever constructed within the package, consumers only ever see the interface.

File: packages/dds/merge-tree/api-report/merge-tree.legacy.alpha.api.md
Comment (anthony-murphy) [inline packages/dds/merge-tree/api-report/merge-tree.legacy.alpha.api.md:91]: @tylerbutler  do you know why these type removals don't break the type tests? i would expect those to need suppression, but they don't for some reason.

Comment (anthony-murphy) [reply packages/dds/merge-tree/api-report/merge-tree.legacy.alpha.api.md:91]: i guess it makes sense if the look at the imports 
```
import type * as old from "@fluidframework/merge-tree-previous/internal";

import type * as current from "../../index.js";
```

as the types still exist off internal, just not legacy, so are still available off `import type * as current from "../../index.js";
`

File: packages/dds/sequence/api-report/sequence.legacy.alpha.api.md
Comment (jason-ha) [inline packages/dds/sequence/api-report/sequence.legacy.alpha.api.md:331]: There are a bunch of changes here that are not to deprecated methods/types.
This includes a very serious type relaxing that means implementation will fail to work properly unless the input matches the more strict expectation.
(The SequenceInterval class overrides IInterval to say that all intervals it works with are SequenceIntervals, which typescript unfortunately allows. This restriction should really be preserved.)

**Review by jason-ha — COMMENTED** [review]
I looked over this more and tried to separate more work that can be done right now as I had been mentioning in other comments. From that I produced [#22948](https://github.com/microsoft/FluidFramework/pull/22948). Have a look.
I also have the remainder of the changes in a stash. Final result is not identical to this as I made some fixes.
I think the ideal process for these overall changes are:
1. #22948 
2. just breaks (blocked on 2.10 opening)
3. more taking advantage of the breaks (change internals)

The healthiest way to maintain 2 & 3 over time is likely to be having branch with two commits and rebasing over main when desired.

File: packages/framework/fluid-framework/src/index.ts
Comment (Abe27342) [inline packages/framework/fluid-framework/src/index.ts:155]: why this change? seems like `SharedString` still needs to be usable as a value in public api, e.g. in dynamic DDS creation:

```const myNewString = await container.create(SharedString)```

Comment (anthony-murphy) [reply packages/framework/fluid-framework/src/index.ts:155]: interesting. i don't think i made an direct modifications to this file, maybe an auto-formatter bug. i'll revert.

Comment (anthony-murphy) [reply packages/framework/fluid-framework/src/index.ts:155]: image

Comment (anthony-murphy) [reply packages/framework/fluid-framework/src/index.ts:155]: this breaks the build. the error doesn't actually call out SharedString. I'll split the import to fix, but odd the autofix does the wrong thing. FYI @Josmithr @tylerbutler

File: packages/dds/sequence/src/test/overlappingSequenceIntervalsIndex.spec.ts
Comment (Abe27342) [inline packages/dds/sequence/src/test/overlappingSequenceIntervalsIndex.spec.ts:23]: nit: can move this import to the index line above, then no need to disable lint rules.

File: packages/framework/fluid-framework/src/index.ts
Comment (anthony-murphy) [reply packages/framework/fluid-framework/src/index.ts:155]: i also don't think we type test fluid-framework, so nothing catches this.

File: packages/dds/sequence/src/intervalIndex/overlappingSequenceIntervalsIndex.ts
Comment (Abe27342) [inline packages/dds/sequence/src/intervalIndex/overlappingSequenceIntervalsIndex.ts:23]: same nit here

File: packages/dds/merge-tree/src/mergeTreeNodes.ts
Comment (Abe27342) [inline packages/dds/merge-tree/src/mergeTreeNodes.ts:267]: nit: rather than remove, might be better to put these in a `\@privateremarks` block? That should make them still visible to merge-tree devs without leaking the info to API surface / docs.

File: packages/framework/fluid-framework/src/index.ts
Comment (Josmithr) [reply packages/framework/fluid-framework/src/index.ts:155]: > this breaks the build. the error doesn't actually call out SharedString. I'll split the import to fix, but odd the autofix does the wrong thing. FYI @Josmithr @tylerbutler

I don't think I've seen the auto-fixer change an export that was not type-only to a type-export before. If it's doing that, we may want to disable auto-fix for that rule.

> i also don't think we type test fluid-framework, so nothing catches this.

This seems like a pretty serious oversight in our tooling. We should schedule work for this.

Comment (anthony-murphy) [reply packages/framework/fluid-framework/src/index.ts:155]: @Josmithr do you know the right place to file a bug like this, and which team would own it? if you do would you mind filing it, and linking it here?

Comment (Josmithr) [reply packages/framework/fluid-framework/src/index.ts:155]: Ownership of some of this tooling stuff remains pretty ambiguous (something my team has been pushing on in retrospectives). I'm happy to take on the eslint update (I have another update I want to make in that space anyways). https://dev.azure.com/fluidframework/internal/_workitems/edit/22618

I think Team Sasha/Jason has been owning the type-test infra? Is that right? @jason-ha

Comment (tylerbutler) [reply packages/framework/fluid-framework/src/index.ts:155]: In the past we haven't worried about type testing fluid-framework since it just re-exports APIs that are type-tested in their "real" packages. But that was a long time ago, and we have since stopped using `export *` so messing up the exports is more likely now. It was also based on the fact that the entire API surface was type-tested, though, and that's no longer the case for most packages which only type-test a subset of their API. It should still be a super-set of what's exposed in fluid-framework, though.

**Review by Abe27342 — APPROVED** [review]

File: .changeset/rude-views-bake.md
Comment (tylerbutler) [inline .changeset/rude-views-bake.md:12]:
```suggestion
The `Client` class in the merge-tree package has been removed. Types that directly or indirectly expose the merge-tree `Client` class have also been removed.
```

**Review by tylerbutler — APPROVED** [review]
Minor nit; docs look good.

**Pull request merged**
